File: src/main/java/eu/karelhovorka/zpevnik/text/SectionTokenizer.java
package eu.karelhovorka.zpevnik.text;


import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static eu.karelhovorka.zpevnik.util.Preconditions.checkNotNull;
import static eu.karelhovorka.zpevnik.util.StringUtil.removeTrailingWhitespace;

public class SectionTokenizer {

    private static final int MINIMUM_SECTION_COUNT = 2;
    private final String splitRegex;

    public static final String TYPE_REGEX = "(?:[A-Z][A-Za-z]*([0-9]*):|([0-9]+)(?:\\.|:))";

    public static final Pattern TYPE_REGEX_PATTERN = Pattern.compile(TYPE_REGEX);

    public static final Pattern VALID_TEXT_PATTERN = Pattern.compile("^" + TYPE_REGEX + "(?:\n|\r).*");

    public SectionTokenizer() {
        //splitRegex = splitRegex();
        splitRegex = "^" + TYPE_REGEX + "$";
    }

    public static boolean isValid(String text) {
        try {
            validate(text);
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    public static void validate(String text) {
        if (!VALID_TEXT_PATTERN.matcher(text).find()) {
            throw new IllegalStateException("Not starting with section header");
        }
        Matcher m = Pattern.compile(TYPE_REGEX + "(?:\n|\r)").matcher(text);
        int count = 0;
        while (m.find()) {
            count += 1;
        }
        if (count < MINIMUM_SECTION_COUNT) {
            throw new IllegalStateException("not enough sections: " + count);
        }
    }

    private String splitRegex() {
        StringBuilder sb = new StringBuilder();
        for (SectionType sectionType : SectionType.values()) {
            for (String name : sectionType.names) {
                sb.append("|" + name + ":");
            }

        }
        return "(^" + sb.substring(1) + "\n)";
    }

    public List<Section> getSections(String text) {
        checkNotNull(text, "text");
        String[] lines = text.split("(\n|\r)+");
        List<Section> sections = new ArrayList<>();
        SectionType st = null;
        StringBuilder sb = null;
        for (String line : lines) {
            if (line.matches(splitRegex)) {
                if (st != null) {
                    sections.add(new Section(removeTrailingWhitespace(sb.toString()), st));
                }
                st = SectionType.fromName(line);
                sb = new StringBuilder();
            } else {
                if (sb == null) {
                    throw new IllegalArgumentException("Text does not start with a section");
                }
                sb.append(line + "\n");
            }
        }
        if (st != null) {
            sections.add(new Section(removeTrailingWhitespace(sb.toString()), st));
        }
        return mergeSections(sections);
    }

    public List<Section> mergeSections(List<Section> originalSections) {
        List<Section> mergedSections = new ArrayList<>();
        for (Section section : originalSections) {
            Section sectionToAdd = section;
            for (int j = mergedSections.size() - 1; j >= 0; j--) {
                Section previousSection = mergedSections.get(j);
                if (section.sameContentAndType(previousSection)) {
                    sectionToAdd = Section.makeCopyOf(previousSection);
                    break;
                }
            }
            mergedSections.add(sectionToAdd);
        }
        return mergedSections;
    }


}

File: src/main/kotlin/eu/karelhovorka/zpevnik/text/SectionTokenizer.kt
package eu.karelhovorka.zpevnik.text


import eu.karelhovorka.zpevnik.text.section.parseSectionText
import mock.*

class SectionTokenizer {
    private val SECTION_TYPE_BASIC_REGEX = "^$TYPE_REGEX$".toRegex()

    private fun splitRegex(): String {
        var sb = ""
        for (sectionType in SectionType.values()) {
            for (name in sectionType.names) {
                sb += ("|$name:")
            }

        }
        return "(^" + sb.substring(1) + "\n)"
    }

    fun getSectionsOrSingleSection(text: String): List<Section> {
        try {
            return getSections(text.trimEnd())
        } catch (e: Exception) {
            //e.printStackTrace()
            return listOf(Section(parseSectionText(text.trimEnd())))
        }
    }

    fun getSections(text: String): List<Section> {
        val lines = text.trimEnd().split("\n".toRegex()).dropLastWhile { it.isEmpty() }
        val sections = ArrayList<Section>()
        var sb: StringBuilder? = null
        val sectionTypeCount: MutableMap<ISectionType, Int> = mutableMapOf()
        var currentHeaderLine = ""
        for (line in lines) {
            if (line.trim().matches(SECTION_TYPE_BASIC_REGEX)) {
                if (currentHeaderLine != "") {
                    sections.add(parseSection(currentHeaderLine, sectionTypeCount, sb))
                }
                currentHeaderLine = line
                sb = StringBuilder()
            } else {
                if (sb == null) {
                    throw IllegalArgumentException("Text does not start with a section, line: '$line'")
                }
                sb.append(line + "\n")
            }
        }
        if (currentHeaderLine != "") {
            sections.add(parseSection(currentHeaderLine, sectionTypeCount, sb))
        }
        return mergeSections(sections)
    }

    private fun parseSection(currentHeaderLine: String, sectionTypeCount: MutableMap<ISectionType, Int>, sb: StringBuilder?): Section {
        val explicitIndex = currentHeaderLine.replace(SECTION_TYPE_BASIC_REGEX, "$1$2")
        val st = SectionType.fromName(currentHeaderLine)
        var index = sectionTypeCount.get(st) ?: 0
        if (explicitIndex.isNotBlank()) {
            try {
                index = explicitIndex.toInt() - 1
            } catch (e: NumberFormatException) {
                if (sb!!.toString().isBlank()) {
                    index = 0;
                }
            }
        } else if (sb!!.toString().isBlank()) {
            index = 0;
        }
        sectionTypeCount.put(st, (sectionTypeCount.get(st) ?: 0) + 1)
        return Section(parseSectionText(sb!!.toString()), st, null, 0, index)
    }

    fun mergeSections(originalSections: List<Section>): List<Section> {
        val mergedSections = ArrayList<Section>()
        for (section in originalSections) {
            var sectionToAdd = section
            for (j in mergedSections.indices.reversed()) {
                val previousSection = mergedSections[j]
                if (section.sameContentAndType(previousSection)) {
                    sectionToAdd = Section.makeCopyOf(previousSection)
                    break
                }
            }
            mergedSections.add(sectionToAdd)
        }
        return mergedSections
    }

    companion object {

        private val MINIMUM_SECTION_COUNT = 2

        @JvmField
        val TYPE_REGEX = "(?:[A-Z][A-Za-z]*([0-9]*):|([0-9]+)\\.)[\t ]*"

        val TYPE_REGEX_PATTERN = TYPE_REGEX.toPattern()

        val VALID_TEXT_PATTERN = "^\\s*$TYPE_REGEX(?:\n|\r).*".toPattern()

        fun isValid(text: String): Boolean {
            try {
                validate(text)
                return true
            } catch (e: Exception) {
                //e.printStackTrace()
                return false
            }

        }

        fun validate(text: String) {
            if (!VALID_TEXT_PATTERN.matcher(text).find()) {
                throw IllegalStateException("Not starting with section header")
            }
            val m = "$TYPE_REGEX(?:\n|\r)".toPattern().matcher(text)
            var count = 0
            while (m.find()) {
                count += 1
            }
            if (count < MINIMUM_SECTION_COUNT) {
                throw IllegalStateException("not enough sections: " + count)
            }
        }
    }


}
